File: Day096_2025/Q096_GFG.java
import java.util.*;
class Q096_GFG {
    public static ArrayList<Integer> topoSort(int V, int[][] edges) {
        ArrayList<Integer>[] adj = new ArrayList[V];
        for (int i = 0; i < V; i++)
            adj[i] = new ArrayList<>();

        for (int[] edge : edges) {
            int u = edge[0], v = edge[1];
            adj[u].add(v);
        }

        int[] inDegree = new int[V];
        for (int u = 0; u < V; u++) {
            for (int v : adj[u])
                inDegree[v]++;
        }

        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < V; i++) {
            if (inDegree[i] == 0)
                q.add(i);
        }

        ArrayList<Integer> result = new ArrayList<>();
        while (!q.isEmpty()) {
            int node = q.poll();
            result.add(node);
            for (int neighbor : adj[node]) {
                inDegree[neighbor]--;
                if (inDegree[neighbor] == 0)
                    q.add(neighbor);
            }
        }

        return result;
    }
}

File: Day096_2025/Q096_GFG.py
class Solution:
    def topoSort(self, V, edges):
        from collections import deque, defaultdict
        
        adj = defaultdict(list)
        inDegree = [0] * V
        
        for u, v in edges:
            adj[u].append(v)
            inDegree[v] += 1

        q = deque([i for i in range(V) if inDegree[i] == 0])
        result = []

        while q:
            node = q.popleft()
            result.append(node)
            for neighbor in adj[node]:
                inDegree[neighbor] -= 1
                if inDegree[neighbor] == 0:
                    q.append(neighbor)

        return result
